src/Moat/Types.hs

```diff
@@ -312,6 +314,12 @@ data Options = Options
     --   The default ('False') will generate newtypes
     --   as their own structs.
     typeAlias :: Bool,
+    -- | Kotlin annotations to add to a newtype.
+    --   The default (@[]@) will add none.
+    --
+    --   If you're targeting the Kotlin/JVM backend,
+    --   it will be useful to add 'JvmInline' here.
+    newtypeAnns :: [Annotation],
```

Comment:
Hmm. Can't we just add `JvmInline` automatically to newtypes?

Reply:
It could be a conditional option?

Reply:
Kotlin has native and JS backends as well, so `JvmInline` isn't needed in all cases. A list of annotations dovetails with `dataAnnotations` and is more flexible for other use-cases as well. So IDK about a specific option for adding `JvmInline` when this covers the use-case.

Reply:
I could make this default to `[JvmInline]` though, would that cover this?

Reply:
It just feels odd to me to have this as a separate field. 

```haskell
{ dataAnnotations = [Parcelize],
  dataInterfaces = [Parcelable],
  dataProtocols = [OtherProtocol "CaseIterable", Hashable, Codable],
  newtypeAnns = [JvmInline]
}
```

This looks weird to me. I don't see why we wouldn't just do,

```haskell
{ dataAnnotations = [Parcelize, JvmInline],
  dataInterfaces = [Parcelable],
  dataProtocols = [OtherProtocol "CaseIterable", Hashable, Codable]
}
```

Alternately, we could add a field `mkNewtypeJvmInline :: Bool` that adds JvmInline to all newtypes and defaults to true.

Reply:
`JvmInline` is only possible on `value class` definitions, so that would cause all `data class` definitions to fail to compile. Having an option for annotating all `value class` annotations would be useful for other purposes; for example, I would like to override `toString` for generated value classes via an annotation processor.

Reply:
Aren't all newtypes, value classes? https://github.com/MercuryTechnologies/moat/blob/b39d4bb5617d12dc4fb2080cfe5fbd439fd8c2ff/src/Moat/Pretty/Kotlin.hs#L35-L45

Reply:
i.e. we wouldn't do anything to non-newtypes with `mkNewtypeJvmInline`

Reply:
I see a two other options here. There are probably more I'm not immediately thinking of.

1. If JvmInline is provided to a non-newtype, we can throw a compile-time error (we are in Q, after all)

2. If JvmInline is provided to a non-newtype, we ignore it when rendering.

I prefer (1). It's pretty simple. In the relevant code for datatypes that are not newtypes:
```haskell
when (JvmInline `elem` dataProtocols) $ do
  throwError NewConstructorYouNeedToAddThatIndicatesTheGoingsWrong
```

This avoids code duplication, ensures the API doesn't bloat unreasonably, and gives the user an informative error message when they do something wrong!

Reply:
MoatError and prettyMoatError are what would need updating.

Reply:
Additionally, I think it's bad to assume what interfaces/protocols users will need - Mercury may very well want to default to newtypes having JvmInline, but others may not. It's easy to define helper functions to do what you need. This is in fact what Mercury does presently.

Reply:
> Additionally, I think it's bad to assume what interfaces/protocols users will need - Mercury may very well want to default to newtypes having JvmInline, but others may not. It's easy to define helper functions to do what you need. This is in fact what Mercury does presently.

I agree, which is why I'm confused about the pushback here; if users want to annotate all newtypes with `@JvmInline`, this is a straightforward option to do so. Am I missing something?

Reply:
My pushback is the addition of `newtypeAnns`. I prefer "1." as chessai outlined.

Reply:
Removed `newtypeAnns` altogether, good discussion!